### PR TITLE
prepare_seq2seq_batch makes labels/ decoder_input_ids made later.

### LINKED ISSUE #6624: Bart: make decoder_input_ids correctly if labels specified.
should call shift_tokens_right(labels), like T5.

cc @patil-suraj: does that make sense?

### LINKED ISSUE #6711: Pegasus finetuning: OOM
Epoch 0:  91% 5747/6331 [39:52<04:03,  2.40it/s, loss=75.765, v_num=2]/usr/local/lib/python3.6/dist-packages/torch/optim/lr_scheduler.py:200: UserWarning: Please also save or load the state of the optimzer when saving or loading the scheduler.
  warnings.warn(SAVE_STATE_WARNING, UserWarning)
tcmalloc: large alloc 1083260928 bytes == 0x1aece0000 @  0x7f144f09c615 0x591f47 0x4cc229 0x4cc38b 0x566c91 0x5a4df1 0x630b1d 0x7f1443355950 0x7f1443359bf7 0x7f144368a7e8 0x7f14436401b3 0x50a47f 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50cfd6 0x509918 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x509918 0x50a64d 0x50c1f4
tcmalloc: large alloc 1354080256 bytes == 0x21e5c000 @  0x7f144f09c615 0x591f47 0x4cc229 0x4cc38b 0x566c91 0x5a4df1 0x630b1d 0x7f1443355950 0x7f1443359bf7 0x7f144368a7e8 0x7f14436401b3 0x50a47f 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50cfd6 0x509918 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x509918 0x50a64d 0x50c1f4
tcmalloc: large alloc 1692606464 bytes == 0x7f10651ce000 @  0x7f144f09c615 0x591f47 0x4cc229 0x4cc38b 0x566c91 0x5a4df1 0x630b1d 0x7f1443355950 0x7f1443359bf7 0x7f144368a7e8 0x7f14436401b3 0x50a47f 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50cfd6 0x509918 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x509918 0x50a64d 0x50c1f4
tcmalloc: large alloc 2115764224 bytes == 0x7f0fe700e000 @  0x7f144f09c615 0x591f47 0x4cc229 0x4cc38b 0x566c91 0x5a4df1 0x630b1d 0x7f1443355950 0x7f1443359bf7 0x7f144368a7e8 0x7f14436401b3 0x50a47f 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50cfd6 0x509918 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x509918 0x50a64d 0x50c1f4
tcmalloc: large alloc 2644705280 bytes == 0x7f0f495de000 @  0x7f144f09c615 0x591f47 0x4cc229 0x4cc38b 0x566c91 0x5a4df1 0x630b1d 0x7f1443355950 0x7f1443359bf7 0x7f144368a7e8 0x7f14436401b3 0x50a47f 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50cfd6 0x509918 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x509918 0x50a64d 0x50c1f4
tcmalloc: large alloc 3305881600 bytes == 0x7f0fe700e000 @  0x7f144f09c615 0x591f47 0x4cc229 0x4cc38b 0x566c91 0x5a4df1 0x630b1d 0x7f1443355950 0x7f1443359bf7 0x7f144368a7e8 0x7f14436401b3 0x50a47f 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50cfd6 0x509918 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x509918 0x50a64d 0x50c1f4
tcmalloc: large alloc 4132356096 bytes == 0x7f0e530f2000 @  0x7f144f09c615 0x591f47 0x4cc229 0x4cc38b 0x566c91 0x5a4df1 0x630b1d 0x7f1443355950 0x7f1443359bf7 0x7f144368a7e8 0x7f14436401b3 0x50a47f 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50cfd6 0x509918 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x509918 0x50a64d 0x50c1f4
tcmalloc: large alloc 5165449216 bytes == 0x7f0f495de000 @  0x7f144f09c615 0x591f47 0x4cc229 0x4cc38b 0x566c91 0x5a4df1 0x630b1d 0x7f1443355950 0x7f1443359bf7 0x7f144368a7e8 0x7f14436401b3 0x50a47f 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50cfd6 0x509918 0x50a64d 0x50c1f4 0x507f24 0x509c50 0x50a64d 0x50c1f4 0x509918 0x50a64d 0x50c1f4
./finetune_pegasus_xsum.sh: line 15:   876 Killed             

I appreciate any help. Thank you.

### PR DESCRIPTION
**Pull request by sshleifer**
`src/` changes:
- when `tgt_texts` is supplied `prepare_seq_to_seq_batch` calls the tensor that used to be called `decoder_input_ids`, `labels`.
- This change helps metrics for models whose tokenizers do not add bos to the beginning of target sequences, like Marian and Pegasus, without affecting metrics for other models (bart). 
This branch was originally called "Fairseq batch equivalence", because it makes batches that look identical to fairseq's for mbart (and bart).
- tokenization testing file for bart.
- lots of cleanup and testing.

`examples/seq2seq` changes:
- `examples/seq2seq/finetune.py` (and eventually Seq2SeqTrainer) makes decoder_input_ids by shifting tokens right.
- this enables Marian finetuning and distillation, with a few extra changes. 
- add `--label_smoothing` option to seq2seq/distillation.py
- rename `Seq2SeqDataset` -> `LegacySeq2SeqDataset` and `TranslationDataset`-> `Seq2SeqDataset`. The new `Seq2SeqDataset` calls `prepare_seq2seq_batch`. The choice of which dataset to use is determined based on whether the tokenizer has a `prepare_seq2seq_batch` method.

**Problem:**

Previously on master, if the target language sequence was
"Șeful ONU declară că nu există soluții militare în Siria", and the tokenizer was Marian, lm_labels would become "ONU declară că nu există soluții militare în Siria", and the model would learn to skip the first token (or not generate bos).

Generations would then start very strangely, for example:
`", fostul şef al personalului prezidenţial din Brazilia, va participa la un proces"`

now: `"Fostul şef al personalului prezidenţial al Braziliei va fi judecat".`
(same thing is happening for pegasus #6711)

### Metrics
**mbart en-> ro**: no change
marian: master: 23 BLEU, this branch: 25
(en ro distillation/no teacher/3 dec layers)
distilbart-cnn-12-3: no change (within 0.01 ROUGE 2)
master + label smoothing: `{'rouge1': 43.2764,  'rouge2': 20.4969, 'rougeL': 29.9210}`
this branch + label smoothing: `{"rouge1": 43.1997, "rouge2": 20.4879, "rougeL": 30.1607}`

### TODO:
- check t5-base
- check pegasus

If you want to test whether this branch makes truncation go away, the easiest way is to pull the mirror branch with
```bash
git fetch
git checkout batch-parity-cleaner
```

cc @patil-suraj

### CODE REVIEW

File: examples/seq2seq/finetune.py [hunk 1]
```diff
@@ -167,7 +180,7 @@ def training_step(self, batch, batch_idx) -> Dict:
 
         logs = {name: loss for name, loss in zip(self.loss_names, loss_tensors)}
         # tokens per batch
-        logs["tpb"] = batch["input_ids"].ne(self.pad).sum() + batch["decoder_input_ids"].ne(self.pad).sum()
+        logs["tpb"] = batch["input_ids"].ne(self.pad).sum() + batch["labels"].ne(self.pad).sum()
```
Comment (sshleifer) [inline examples/seq2seq/finetune.py:183]: clearly we dont have any test coverage for LegacyDataset!

**Review by LysandreJik — COMMENTED** [review]
Love the variable renaming/readability changes.

Can the BART tokenization tests leverage the `TokenizerTesterMixin`?

File: examples/seq2seq/distillation.py [hunk 2]
```diff
@@ -246,7 +260,7 @@ def add_distill_args(parser):
 
 class BartTranslationDistiller(BartSummarizationDistiller):
     mode = "translation"
-    loss_names = ["loss"]
+    loss_names = ["loss", "ce_loss", "mlm_loss", "enc_mse_loss", "hid_loss_enc", "hid_loss_dec"]
```
Comment (sshleifer) [inline examples/seq2seq/distillation.py:263]: QOL improvement unrelated to the batching change

File: examples/seq2seq/distillation.py [hunk 3]
```diff
@@ -424,33 +439,40 @@ def evaluate_checkpoint(ckpt_path: Path, dest_dir=None):
...
-            4: [0, 5, 10, 15],
-            6: [0, 3, 6, 9, 12, 15],
-            8: [0, 2, 4, 6, 8, 10, 12, 15],
-            9: [0, 1, 3, 5, 7, 9, 11, 13, 15],
-            16: all_layers,
-        }
-        return layers_to_copy[n_to_get]
-    else:
-        return all_layers[:n_to_get]  # TODO: better version on theseus-bart branch
+LAYERS_TO_COPY = {
```
Comment (sshleifer) [inline examples/seq2seq/distillation.py:442]: QOL improvement unrelated to the batching change

File: examples/seq2seq/finetune.py [hunk 4]
```diff
@@ -102,14 +103,13 @@ def __init__(self, hparams, **kwargs):
...
-        self.decoder_start_token_id = None
+        self.decoder_start_token_id = None  # default to config
         if self.model.config.decoder_start_token_id is None and isinstance(self.tokenizer, MBartTokenizer):
             self.decoder_start_token_id = self.tokenizer.lang_code_to_id[hparams.tgt_lang]
             self.model.config.decoder_start_token_id = self.decoder_start_token_id
-        if isinstance(self.tokenizer, MBartTokenizer) or isinstance(self.tokenizer, MarianTokenizer):
-            self.dataset_class = TranslationDataset
-        else:
-            self.dataset_class = Seq2SeqDataset
+        self.dataset_class = (
```
Comment (sshleifer) [inline examples/seq2seq/finetune.py:110]: important change

File: examples/seq2seq/finetune.py [hunk 5]
```diff
@@ -134,19 +134,33 @@ def ids_to_clean_text(self, generated_ids: List[int]):
...
-        source_ids, source_mask, target_ids = batch["input_ids"], batch["attention_mask"], batch["decoder_input_ids"]
+        source_ids, source_mask = batch["input_ids"], batch["attention_mask"]
+        decoder_attention_mask = None  # bart will make this, we supply for t5
 
         if isinstance(self.model, T5ForConditionalGeneration):
-            decoder_input_ids = self.model._shift_right(target_ids)
-            lm_labels = target_ids
+            lm_labels = batch["labels"]
+            decoder_input_ids = self.model._shift_right(lm_labels)
+            decoder_attention_mask = decoder_input_ids.ne(pad_token_id)
```
Comment (sshleifer) [inline examples/seq2seq/finetune.py:143]: this is new

File: examples/seq2seq/finetune.py [hunk 6]
```diff
@@ -134,19 +134,33 @@ def ids_to_clean_text(self, generated_ids: List[int]):
...
+        elif "labels" in batch:
+            lm_labels = batch["labels"]
+            decoder_input_ids = shift_tokens_right(lm_labels, pad_token_id)
```
Comment (sshleifer) [inline examples/seq2seq/finetune.py:146]: This is the key change

File: examples/seq2seq/finetune.py [hunk 7]
```diff
@@ -134,19 +134,33 @@ def ids_to_clean_text(self, generated_ids: List[int]):
...
         else:
-            decoder_input_ids = target_ids[:, :-1].contiguous()  # Why this line?
-            lm_labels = target_ids[:, 1:].clone()  # why clone?
-
-        outputs = self(source_ids, attention_mask=source_mask, decoder_input_ids=decoder_input_ids, use_cache=False)
+            # TODO: fix inside LegacyDataset
+            target_ids = batch["decoder_input_ids"]
```
Comment (sshleifer) [inline examples/seq2seq/finetune.py:149]: fall back on old shifting logic

File: examples/seq2seq/finetune.py (hunk 4 above)
Comment (patil-suraj) [reply examples/seq2seq/finetune.py:110]: If every seq2seq model is going to have the `prepare_seq2seq_batch` then we might not need `LegacyDataset`. Would it be a good idea to make make `prepare_seq2seq_batch` batch mandatory for all seq2seq models and not maintain two datasets ?

Comment (sshleifer) [reply examples/seq2seq/finetune.py:110]: Maybe eventually. @patrickvonplaten is experimenting with composing GLUE models into seq2seq models with `EncoderDecoderModel`, (e.g. `Roberta2Roberta`), so I think that we keep the Legacy dataset until we can guarantee that most tokenizers have `prepare_seq2seq_batch` methods, which might be never.

**Comment by sshleifer** [comment]
Added common tokenizer tests @LysandreJik

**Review by LysandreJik — APPROVED** [review]
LGTM, cool tests!
